### PR TITLE
Coordinate Representation Classes

### PR DESCRIPTION
**Pull request by Cadair**
Have a very merry Friday.

### CODE REVIEW

**Comment by Cadair** [comment]
ping @astrofrog

**Comment by astrofrog** [comment]
@eteq - there are a few changes compared to the API document:
- `PhysicistSphericalRepresentation` is `PhysicsSphericalRepresentation`
- `representation=...` is now a class method instead, simplifying the code a lot
- We've added the unitspherical representation

Transformations are implemented in a simple way - by default, all transformations go through cartesian, but one can add short cuts by overloading `represent_as`. It's simple, but it works well. We still need to overload some of the `represent_as` methods for e.g. `UnitSpherical` but that's just optimization, and doesn't change the functionality.

Also, what exactly is the physics spherical representation? At the moment, it simply renames `lon` to `phi` and `lat` to `theta`, but should `theta = 90 - lat`?

**Comment by eteq** [comment]
To answer the last question: Yep, it's just what you're thinking: theta goes from the z-axis to the x-axis (of phi=0) - a la http://en.wikipedia.org/wiki/File:3D_Spherical.svg

File: astropy/coordinates/representation.py [hunk 1]
```diff
@@ -0,0 +1,540 @@
...
+    """
+
+    def __init__(self, x, y=None, z=None, unit=None, copy=True):
+
+        if y is None and z is None:
+            x, y, z = x
+        elif (y is None and z is not None) or (y is not None and z is None):
+            raise ValueError("x, y, and z are required to instantiate CartesianRepresentation")
+
+        if unit is not None:
```
Comment (mhvk) [inline astropy/coordinates/representation.py:102]: I would change this to:

```
if unit is None:
    unit = x.unit
else:
    unit = u.Unit(unit)
```

The logic is that if no unit is given, the input values have to be quantities. Furthermore, with this change, the next few lines will automatically get all the units on the same scale, and thus you can remove the (slow) `physical_type` check.

Comment (mhvk) [inline astropy/coordinates/representation.py:102]: p.s.  Maybe with a try/except, like:

```
    try:
        unit = x.unit
    except AttributeError:
        raise TypeError("Input coordinates should be Quantities with units "
                        "if no explicit unit is given")
```

File: astropy/coordinates/representation.py [hunk 2]
```diff
@@ -0,0 +1,540 @@
...
+            unit = u.Unit(unit)
+
+        x = u.Quantity(x, unit=unit, copy=copy)
```
Comment (mhvk) [inline astropy/coordinates/representation.py:105]: If you do the above, you could also have a try/except around these three lines, raising the `UnitsError` from below if it fails.

File: astropy/coordinates/representation.py [hunk 3]
```diff
@@ -0,0 +1,540 @@
...
+            raise ValueError("Input parameters x, y, and z cannot be broadcast")
+
+        self._x = x
+        self._y = y
+        self._z = z
+
+    @property
+    def x(self):
+        """
+        The x position of the point(s).
```
Comment (mhvk) [inline astropy/coordinates/representation.py]: `position` -> `component` (since these do not have to be lengths any more...)

File: astropy/coordinates/representation.py [hunk 4]
```diff
@@ -0,0 +1,540 @@
...
+    Parameters
+    ----------
+    x, y, z : `~astropy.units.Quantity` or float or `~numpy.ndarray`
+        The x, y, and z coordinates of the point(s), which should either be
+        `~astropy.units.Quantity` instances, or can be passed as
+        float or `numpy.ndarray` provided that the ``unit`` parameter is
+        specified. If ``x``, ``y``, and ``z`` have different shapes, they
+        should be broadcastable.
+
+    unit : `~astropy.units.Unit`, optional
```
Comment (mhvk) [inline astropy/coordinates/representation.py:84]: Just to be sure, do we really want the option to give the unit? It makes the class somewhat asymmetric with respect to Spherical and Cylindrical.

File: astropy/coordinates/representation.py [hunk 5]
```diff
@@ -0,0 +1,540 @@
...
+        If True arrays will be copied rather than referenced.
+    """
+
+    def __init__(self, lon, lat, distance, copy=True):
+
+        # Let the Longitude and Latitude classes deal with e.g. parsing
+        lon = Longitude(lon, copy=copy)
+        lat = Latitude(lat, copy=copy)
+
+        if isinstance(distance, u.Quantity) and distance.unit.physical_type == 'length':
```
Comment (mhvk) [inline astropy/coordinates/representation.py]: For this class, `distance` has to be a `Quantity`, correct? If so,

```
if isinstance(distance, u.Quantity):
    if distance.unit.physical_type ...
else:
    raise TypeError("Distance should be a Quantity").
```

Alternatively, have a `unit` argument, in analogy with `CartesianRepresentation`
(though it would be a bit "unfair" that the angles would have to have their own units).

File: astropy/coordinates/representation.py [hunk 6]
```diff
@@ -0,0 +1,540 @@
...
+        else:
+            return super(SphericalRepresentation, self).represent_as(other_class)
+
+    def to_cartesian(self):
+        """
+        Converts spherical polar coordinates to 3D rectangular cartesian
+        coordinates.
+        """
+
+        # We need to convert Distance to Quantity to allow negative values.
```
Comment (mhvk) [inline astropy/coordinates/representation.py:240]: Why is the `view` not an easy way? It will remain the fastest! (and with the `subok` option, that is what `Quantity` does inside anyway).

Comment (astrofrog) [inline astropy/coordinates/representation.py:240]: It is - the comment refers to the previous way I had done it... oops!

File: astropy/coordinates/representation.py [hunk 7]
```diff
@@ -0,0 +1,540 @@
...
+        return CartesianRepresentation(x=x, y=y, z=z)
+
+    @classmethod
+    def from_cartesian(cls, cart):
+        """
+        Converts 3D rectangular cartesian coordinates to spherical polar
+        coordinates.
+        """
+
+        xsq = cart.x ** 2
```
Comment (mhvk) [inline astropy/coordinates/representation.py]: Could replace this with

```
s = np.hypot(cart.x, cart.y)
r = np.hypot(s, cart.x)
```

I think this will be (much) faster despite the extra squaring since it will avoid creating new objects/storing in memory.

File: astropy/coordinates/representation.py [hunk 8]
```diff
@@ -0,0 +1,540 @@
...
```
Comment (mhvk) [inline astropy/coordinates/representation.py]: As above

```
s = np.hypot(cart.x, cart.z)
```

File: astropy/coordinates/representation.py [hunk 9]
```diff
@@ -0,0 +1,540 @@
...
```
Comment (mhvk) [inline astropy/coordinates/representation.py]: As above.

File: astropy/coordinates/representation.py (hunk 7 above)
Comment (astrofrog) [inline astropy/coordinates/representation.py]: `np.hypot`?? You learn something every day!

File: astropy/coordinates/representation.py [hunk 10]
```diff
@@ -0,0 +1,540 @@
...
+        else:
+            return super(PhysicsSphericalRepresentation, self).represent_as(other_class)
+
+    def to_cartesian(self):
+        """
+        Converts spherical polar coordinates to 3D rectangular cartesian
+        coordinates.
+        """
+
+        # We need to convert Distance to Quantity to allow negative values.
```
Comment (mhvk) [inline astropy/coordinates/representation.py:429]: As above.

File: astropy/coordinates/representation.py [hunk 11]
```diff
@@ -0,0 +1,540 @@
...
+
+    copy : bool, optional
+        If True arrays will be copied rather than referenced.
+    """
+
+    def __init__(self, rho, phi, z, copy=True):
+
+        rho = u.Quantity(rho, copy=copy)
+        phi = Angle(phi, copy=copy)
+        z = u.Quantity(z, copy=copy)
```
Comment (mhvk) [inline astropy/coordinates/representation.py:482]: I'd do

```
try:
    z = u.Quantity(z, rho.unit, copy=copy)
except u.UnitsError:
    raise u.UnitsError("rho and z ...")
```

File: astropy/coordinates/representation.py [hunk 12]
```diff
@@ -0,0 +1,540 @@
...
+        return self._z
+
+    @classmethod
+    def from_cartesian(cls, cart):
+        """
+        Converts 3D rectangular cartesian coordinates to cylindrical polar
+        coordinates.
+        """
+
+        rho = np.sqrt(cart.x ** 2 + cart.y ** 2)
```
Comment (mhvk) [inline astropy/coordinates/representation.py]:
```
rho = np.hypot(cart.x, cart.y)
```

File: astropy/coordinates/representation.py [hunk 13]
```diff
@@ -0,0 +1,540 @@
...
+
+        phi = np.zeros(cart.x.shape) * u.deg
```
Comment (mhvk) [inline astropy/coordinates/representation.py]: Why is this necessary? `np.arctan2(np.zeros(1),np.zeros(1))` seems to work fine. If there are python/numpy versions for which this is not the case, maybe change error reporting temporarily?

File: astropy/coordinates/representation.py [hunk 14]
```diff
@@ -0,0 +1,541 @@
+# In this file, we define the coordinate representation classes, which are
```
Comment (eteq) [inline astropy/coordinates/representation.py]: This should be a docstring instead of a comment form.  This module may not necessarily end up itself part of the public API, but I think we probably should to be safe and guide anyone in ipython who looks at the module (because it will be present there even if it isn't officially in the API).

Also, minor comment: to be consistent with other docs this should be "in this module", not "in this file" if its a docstring.

File: astropy/coordinates/representation.py [hunk 15]
```diff
@@ -0,0 +1,541 @@
...
+from .angles import Angle, Longitude, Latitude
+from .distances import Distance
+from ..extern import six
+
+# Suggestions to improve API
+#
+# - change PhysicistSphericalRepresentation to PhysicsSphericalRepresentation
+# (implemented below).
+#
+# - add a rotated() method that can rotate the 3D coordinates
```
Comment (eteq) [inline astropy/coordinates/representation.py]: `rotated` should _not_ be in here, in my opinion.  This is implemented as part of the transformations, right now, which is better because it lets you compose them.  Adding it here too creates two ways to do the same thing.

(So I'd say just remove this "Suggestions to improve API" section now, because the other one is already implemented)

File: astropy/coordinates/representation.py [hunk 16]
```diff
@@ -0,0 +1,541 @@
...
+
+__all__ = ["CartesianRepresentation", "SphericalRepresentation",
+           "UnitSphericalRepresentation", "PhysicsSphericalRepresentation",
+           "CylindricalRepresentation"]
+
+
+def broadcast_quantity(*args, **kwargs):
```
Comment (eteq) [inline astropy/coordinates/representation.py:24]: It seems to me that this is actually generally enough to perhaps belong in `astropy.units`.  For now its fine to be here so we can move forward, but perhaps issue a PR to put it in there, and then later we can change this if its accepted?

File: astropy/coordinates/representation.py [hunk 17]
```diff
@@ -0,0 +1,541 @@
...
+    """
+    new_arrays = np.broadcast_arrays(*args)
+    new_quantities = []
+    for i in range(len(new_arrays)):
+        new_quantities.append(args[i].__class__(new_arrays[i], unit=args[i].unit, **kwargs))
+    return tuple(new_quantities)
+
+
+@six.add_metaclass(abc.ABCMeta)
+class BaseRepresentation(object):
```
Comment (eteq) [inline astropy/coordinates/representation.py:36]: This class needs to be part of the public API, so that users/other package devs can subclass it.  So that means it needs to be in `__all__`.  Also, with that change, I'd strongly suggest adding a "Notes" section that describes exactly what subclasses should override.  We'll probably need to add more about this later in the actual docs, but for now it should at least be in the docstrings.

File: astropy/coordinates/representation.py (hunk 4 above)
Comment (eteq) [inline astropy/coordinates/representation.py:84]: @mhvk - you mean that instead the unit should be _inferred_ from `x` `y` and `z` being quantities, right instead of being given explicitly?  I agree with that, unless @Cadair or @astrofrog have objections.

File: astropy/coordinates/representation.py [hunk 18]
```diff
@@ -0,0 +1,541 @@
...
+
+class SphericalRepresentation(BaseRepresentation):
+    """
+    Representation of points in 3D spherical coordinates.
+
+    Parameters
+    ----------
+    lon, lat : `~astropy.units.Quantity` or str
+        The longitude and latitude of the point(s). The input values are
+        passed to the `~astropy.coordinates.Longitude` and
```
Comment (eteq) [inline astropy/coordinates/representation.py:165]: Minor re-word suggestion: instead of "passed to the", it should say "this is a".  From the users perspective, these are the _attributes_ of the class as much as they are the parameters of the initializer. (And its a bit awkward to duplicate this in an "attributes" section for something this small.

File: astropy/coordinates/representation.py [hunk 19]
```diff
@@ -0,0 +1,541 @@
...
+        `~astropy.coordinates.Latitude` class respectively, so any valid
+        input for these classes is acceptable. This includes
+        `~astropy.units.Quantity` instances, strings, lists of strings, and
+        so on. `~astropy.coordinates.Longitude` instances can only be passed
+        to ``lon``, and `~astropy.coordinates.Latitude` instances can only be
+        passed to ``lat``.
+
+    distance : `~astropy.units.Quantity`
+        The distance to the point(s). If the distance is a length, it is
+        passed to the :class:`~astropy.coordinates.Distance` class, otherwise
```
Comment (eteq) [inline astropy/coordinates/representation.py:175]: sams as `lon` `lat` above.

**Comment by eteq** [comment]
@Cadair or @astrofrog - are you planning on addressing the comments soon, or are you waiting for more input (I just gave some)?  Once the comments here are addressed, I'm fine with merging...

File: astropy/coordinates/representation.py (hunk 16 above)
Comment (mhvk) [inline astropy/coordinates/representation.py:24]: That might go in https://github.com/astropy/astropy/issues/1274

File: astropy/coordinates/representation.py (hunk 4 above)
Comment (mhvk) [inline astropy/coordinates/representation.py:84]: Yes, I think there is no need to give an explicit unit; it can be inferred from x,y,z, as it is in the other representations.

(My comment about symmetry implies that if we do stick with the option of giving an explicit unit, I think we should do the same for the other representations.)

Comment (Cadair) [inline astropy/coordinates/representation.py:84]: I am pretty sure we implemented this from the API doc. It allows the shortcut:

``` Python
CartesianRepresentation(10,10,10,unit=u.km)
```

File: astropy/coordinates/representation.py (hunk 1 above)
Comment (Cadair) [inline astropy/coordinates/representation.py:102]: If I understand this correctly, this would force all the coords to be the same unit?

File: astropy/coordinates/representation.py (hunk 7 above)
Comment (Cadair) [inline astropy/coordinates/representation.py]: Nice! Shame it doesn't allow for 3 coordinates.

File: astropy/coordinates/representation.py (hunk 11 above)
Comment (Cadair) [inline astropy/coordinates/representation.py:482]: As above, surely we do not want to force the different coordinates to be in the same units, merely compatible types.

File: astropy/coordinates/representation.py (hunk 1 above)
Comment (mhvk) [inline astropy/coordinates/representation.py:102]: Yes, once `unit` is set to something else than `None`, the `u.Quantity` initilializer will convert whatever unit `x`, `y`, and `z` came with to `unit` (and raise an exception if this is not possible; which is why you would be able to omit the physical type check; of course, if you wish, you could still do the `u.Quantity` inside another try/except so that you can catch the `u.UnitsError` and change the exception text; my sense, though, is that this would be superfluous).

File: astropy/coordinates/representation.py (hunk 11 above)
Comment (mhvk) [inline astropy/coordinates/representation.py:482]: This call will just convert the unit if it is compatible, so in that sense it is no different from what you have. The one difference is that z will indeed be converted to `rho.unit`. To me, that seems desirable (although not essential; it does mean the unit will roundtrip via `CartesianRepresentation`).

File: astropy/coordinates/representation.py (hunk 16 above)
Comment (astrofrog) [inline astropy/coordinates/representation.py:24]: Yes, that was what prompted my question for astropy/astropy#1274

**Comment by astrofrog** [comment]
@eteq - I'd be happy to address the comments today or tomorrow, unless @Cadair is already planning on doing so.

**Comment by Cadair** [comment]
@astrofrog I am rather busy at the moment. So I will leave it to you unless I find some time tonight.

**Comment by mhvk** [comment]
While working on `EarthLocation`, I realised one thing that still would be missing: getting items and slicing, i.e., a definition of `__getitem__`. Furthermore, should there be a basic `__str__` and `__repr__`?

I do wonder a little whether it would be better to add those explicitly, or perhaps subclass from something that has implemented it already: we will probably get more "containers" consisting of quantities with the same shape but possibly different units. And of course have a relevant class: `Table` (at least once `Column` can hold `Quantity`; I really should try to get that working!).

**Comment by eteq** [comment]
@mhvk - good point on slicing - probably we only want to accept "1d" slicing (either in `bool` or list of integers form), and yield the same kind of object as whatever it starts as.  Anything more complicated I think it's fair to just tell the user to use `rep.x` or `rep.lat` or whatever.

As for `__str__` and `__repr__`, That could be here, or we could punt that to the frames.  They'll have to implement their own version of it, anyway, to get the names right (e.g., they should say things like `<ICRS ra=12.1 deg dec=13.5 deg distance=200 kpc>`, and the `ra` and `dec` names are unique to the frame).  So @astrofrog and/or @Cadair, I'm fine with whatever you prefer to do on a short-term basis, and we can always add that later if it makes sense to do.

**Comment by mhvk** [comment]
To all following this: I tried to make a frame work in the context of implementing a subclass-aware version of `np.broadcast_arrays`; see #2327

File: astropy/coordinates/representation.py (hunk 7 above)
Comment (mhvk) [inline astropy/coordinates/representation.py]: @Cadair, @astrofrog - happened to notice that one can do the equivalent of `np.hypot` for arbitrary length, using `np.linalg.norm` [1]. Not that you need it here...
[1] http://docs.scipy.org/doc/numpy/reference/generated/numpy.linalg.norm.html#numpy.linalg.norm

Comment (eteq) [inline astropy/coordinates/representation.py]: @astrofrog  @Cadair @mhvk I use `np.hypot` all the time, but `np.linalg.norm` is new to me.  Unfortunately, it seems it's a lot slower for the 2d case:

```
In [40]: a=randn(10000)
In [41]: b=randn(10000)
In [42]: timeit np.hypot(a,b)
10000 loops, best of 3: 52.5 µs per loop
In [43]: timeit np.linalg.norm([a,b],axis=0)
1000 loops, best of 3: 1.48 ms per loop
In [44]: timeit (a**2+b**2)**0.5
10000 loops, best of 3: 40.2 µs per loop
```

weirdly, it seems `(a**2+b**2)**0.5` is the winner in terms of speed (although close to `hypot`).

**Comment by astrofrog** [comment]
@mhvk - thanks!

I'm going to work on implementing the comments now.

File: astropy/coordinates/representation.py (hunk 1 above)
Comment (astrofrog) [inline astropy/coordinates/representation.py:102]: This does not work in the case where `x=[3 * u.m, 4 * u.km]` since it is neither a quantity nor is `unit` set explicitly. This started to get very messy code-wise so I'm going to leave as-is for now, and we can worry about it if this becomes a bottleneck.

File: astropy/coordinates/representation.py (hunk 3 above)
Comment (astrofrog) [inline astropy/coordinates/representation.py]: Fixed

File: astropy/coordinates/representation.py (hunk 4 above)
Comment (astrofrog) [inline astropy/coordinates/representation.py:84]: @eteq @mhvk - yes, this was straight from the API document. However, I'm perfectly happy to drop this if we think it is inconsistent with other classes. Let us know what you think.

File: astropy/coordinates/representation.py (hunk 5 above)
Comment (astrofrog) [inline astropy/coordinates/representation.py]: I think the API allows `distance=[1.*u.kpc, 500*u.pc]` which is not a Quantity, hence why the code is the current way.

File: astropy/coordinates/representation.py (hunk 6 above)
Comment (astrofrog) [inline astropy/coordinates/representation.py:240]: Fixed

File: astropy/coordinates/representation.py (hunk 7 above)
Comment (astrofrog) [inline astropy/coordinates/representation.py]: Fixed

File: astropy/coordinates/representation.py (hunk 11 above)
Comment (astrofrog) [inline astropy/coordinates/representation.py:482]: I'm not sure if we should change this since it's not clear to me whether the unit of `z` should match that of `rho` in general. I'd prefer to punt on this and implement if we feel the need to enforce this.

File: astropy/coordinates/representation.py (hunk 13 above)
Comment (astrofrog) [inline astropy/coordinates/representation.py]: Fixed

File: astropy/coordinates/representation.py (hunk 14 above)
Comment (astrofrog) [inline astropy/coordinates/representation.py]: Fixed

File: astropy/coordinates/representation.py (hunk 15 above)
Comment (astrofrog) [inline astropy/coordinates/representation.py]: Done

File: astropy/coordinates/representation.py (hunk 17 above)
Comment (astrofrog) [inline astropy/coordinates/representation.py:36]: Done

File: astropy/coordinates/representation.py (hunk 18 above)
Comment (astrofrog) [inline astropy/coordinates/representation.py:165]: I'm not sure how to rephrase it while still mentioning that the user should look at the documentation of Longitude and Latitude for more information on acceptable inputs - any suggestions?

**Comment by astrofrog** [comment]
@Cadair - my fixes are in https://github.com/Cadair/astropy/pull/1

**Comment by astrofrog** [comment]
I'm going to punt on the `__getitem__` for now. I suggest we try and get this in already and then open a new pull request to add other features such as slicing.

**Comment by eteq** [comment]
@astrofrog - I'm fine on punting on `__getitem__` in this PR, but are you fine with the API plan of including this?  Its relevant because if so, I can just delegate `__geitem__` from the frame classes to the representations.  (And presumable the high-level classes can do the same to the frame classes, @taldcroft)

**Comment by astrofrog** [comment]
@eteq - absolutely, I think slicing should work. Do we want to allow scalar representation objects, and if so, should slicing just return an exception?

**Comment by eteq** [comment]
@astrofrog - yeah, I would say so... `CartesianRepresentation(1,2,3)` should be different from `CartesianRepresentation([1], [2], [3])`, after all.  I _think_ you can get away with all of the representations doing something like:

```
def __getitem__(self, slc):
    return self.__class__(self.x[slc], self.y[slc], self.z[slc])
```

or similar, and then let the initializer decide if that's valid?  But maybe its more complicated then that...

**Comment by astrofrog** [comment]
@Cadair - see https://github.com/Cadair/astropy/pull/2 for some fixes and implementation of `__getitem__` - it was as simple as @eteq pointed out.

File: astropy/coordinates/representation.py (hunk 5 above)
Comment (mhvk) [inline astropy/coordinates/representation.py]: But for that example, you now get a `Quantity` instead of a distance. How about

```
    distance = u.Quantity(distance)
    if distance.unit.physical_type == 'length':
        distance = distance.view(Distance)
```

File: astropy/coordinates/representation.py (hunk 11 above)
Comment (mhvk) [inline astropy/coordinates/representation.py:482]: I guess one can convert to cartesian without the physical type matching, but not to spherical. My own sense would be to be strict here; it is really a different beast if the unit of `z` does not match that of `rho` (in physical type at least).

Comment (astrofrog) [inline astropy/coordinates/representation.py:482]: We do currently check that the physical types are the same - it's just the exact units that can be different. Does that seem ok to you?

**Comment by mhvk** [comment]
@astrofrog - agree to keep it simple for now. My guess would be that in the end we want an underlying vector class, which also allows addition, etc. 

I was also thinking of writing a quick `__str__` and `__repr__`; where would I send a PR?

**Comment by Cadair** [comment]
@mhvk send a pull request to Cadair/astropy:ape5-repr

**Comment by eteq** [comment]
@mhvk @astrofrog - I'm actually marginally against implementing `__add__` and such because it won't always mean the same thing for different representations (for Cartesian it might make sense, but everything else is ambiguous).  But I agree we can decide that later.  

`__getitem__` does need to be in for 0.4, though, if we want to have the catalog-matching stuff in the high-level classe.  That's pretty useless without 1D indexing (which means we also need it in the frame and representation classes).

EDIT: Oh, I see @astrofrog added that already - awesome!

**Comment by astrofrog** [comment]
I think I'm missing something, but did we ever discuss adding `__add__`? I agree it should not be added for now.

**Comment by eteq** [comment]
@astrofrog - thats what I thought @mhvk was saying in his last comment?  And are you done making changes now, @astrofrog and @Cadair ?  If so, I can merge this and @mhvk can issue a PR against eteq/astropy:coordinates-ape5 .  That will make it easier to start hooking this into the frames stuff.

**Comment by astrofrog** [comment]
Ah I see - sorry for the confusion. I need to make one more commit to address a comment @mhvk made, then we are good to go.

**Comment by astrofrog** [comment]
@Cadair - can you merge in https://github.com/Cadair/astropy/pull/3

File: astropy/coordinates/representation.py (hunk 11 above)
Comment (mhvk) [inline astropy/coordinates/representation.py:482]: It was mostly that you could remove the check below if you follow my suggestion. Not very important, though.

**Comment by mhvk** [comment]
No worry about `__add__` for now. I'll make a PR. My sense would be that `__mul__` is more important (add distances). Anyway, all for later.

File: astropy/coordinates/representation.py (hunk 7 above)
Comment (mhvk) [inline astropy/coordinates/representation.py]: > weirdly, it seems `(a**2+b**2)**0.5` is the winner in terms of speed (although close to `hypot`).

For large arrays, I get that `np.hypot` is a factor two slower than the
direct calculation, which is rather sad!  Clearly, it is not optimized
properly.  May want to change back (eventually).

**Comment by astrofrog** [comment]
@eteq - this is good to merge!

**Comment by eteq** [comment]
Alright, merging now! @mhvk - you can issue your `__str__` and `__repr__` to my eteq/astropy:coordinates-ape5 , then.
